File: solutions/elixir/bob/1/lib/bob.ex
defmodule Bob do
  @spec hey(String.t()) :: String.t()
  def hey(input) do
    trimmed_input = String.trim(input)
    case {String.ends_with?(trimmed_input, "?"), input === String.upcase(input) and String.downcase(input) !== input, trimmed_input} do
      {true, true, _} -> "Calm down, I know what I'm doing!"
      {true, _, _} -> "Sure."
      {_, _, ""} -> "Fine. Be that way!"
      {_, true, _} -> "Whoa, chill out!"
      _ -> "Whatever."
    end
  end
end

File: solutions/erlang/bob/1/src/bob.erl
-module(bob).

-export([response/1]).

response(Sentence) -> 
    TrimmedSentence = string:trim(Sentence),
    IsYelling = isYelling(TrimmedSentence),
    IsQuestion = isQuestion(TrimmedSentence),
    IsSilence = isSilence(TrimmedSentence),
    if 
      IsYelling and IsQuestion -> "Calm down, I know what I'm doing!";
      IsQuestion -> "Sure.";
      IsYelling -> "Whoa, chill out!";
      IsSilence -> "Fine. Be that way!";
      true -> "Whatever."
   end.

isYelling(Sentence) -> (string:uppercase(Sentence) == Sentence) and (string:lowercase(Sentence) /= Sentence).

isQuestion("") -> false;
isQuestion(Sentence) -> hd(string:reverse(Sentence)) == $?.

isSilence("") -> true;
isSilence(_) -> false.